File: python/packages/spore_tools/src/spore_tools/tools/todo.py
"""TodoWrite tool (#81, net-new Tier-2 storage tool).

Mirrors ``rust/crates/spore-core/src/tools/todo.rs``.

``todo_write`` persists an agent-managed todo list via the :class:`ToolContext`'s
:class:`~spore_core.storage.RunStore` under the key :data:`TODO_STORE_KEY`
(``"todo"``), keyed by the run's :class:`~spore_core.harness.SessionId`. The
agent supplies the FULL desired list on every call; it REPLACES the persisted
list wholesale (no per-item diffing). The current list is returned as JSON
success content.

Like :class:`~spore_tools.tools.tasklist.TaskListTool` it is NOT annotated
``read_only``: it mutates shared persisted state and must dispatch sequentially
(a concurrent read-modify-write would race).
"""

from __future__ import annotations

import json

from spore_core.harness import (
    SandboxProvider,
    ToolOutput,
    ToolOutputError,
    ToolOutputSuccess,
)
from spore_core.model import ToolCall
from spore_core.storage import StorageError
from spore_core.tool_registry import ToolAnnotations, ToolContext, ToolSchema

from .error import ToolExecutionError
from .params import TodoWriteParams, parse_params

#: RunStore key under which the todo list is persisted (issue #81, Q5).
TODO_STORE_KEY = "todo"


class TodoWriteTool:
    NAME = "todo_write"

    def name(self) -> str:
        return self.NAME

    def is_subagent_tool(self) -> bool:
        return False

    def may_produce_large_output(self) -> bool:
        return False

    @classmethod
    def schema(cls) -> ToolSchema:
        return ToolSchema(
            name=cls.NAME,
            description="Replace the persisted todo list with the supplied full list",
            parameters={
                "type": "object",
                "properties": {
                    "todos": {
                        "type": "array",
                        "items": {
                            "type": "object",
                            "properties": {
                                "content": {"type": "string"},
                                "status": {
                                    "type": "string",
                                    "enum": ["completed", "in_progress", "pending"],
                                },
                            },
                            "required": ["content", "status"],
                        },
                    },
                },
                "required": ["todos"],
            },
            # Intentionally NOT read_only — mutates shared persisted state and
            # must dispatch sequentially. See module docs / TaskListTool.
            annotations=ToolAnnotations(),
        )

    async def execute(
        self, call: ToolCall, sandbox: SandboxProvider, ctx: ToolContext
    ) -> ToolOutput:
        try:
            params = parse_params(TodoWriteParams, call)
        except ToolExecutionError as e:
            return e.to_tool_output()

        # Serialize the todos to plain JSON-compatible dicts for both the store
        # and the returned content.
        value = [item.model_dump(mode="json") for item in params.todos]
        try:
            await ctx.run_store.put(ctx.session_id, TODO_STORE_KEY, value)
        except StorageError as e:
            return ToolOutputError(message=f"could not persist todos: {e}", recoverable=True)
        return ToolOutputSuccess(content=json.dumps(value), truncated=False)


__all__ = ["TODO_STORE_KEY", "TodoWriteTool"]

File: typescript/packages/tools/src/todo.ts
/**
 * TodoWrite tool (#81, net-new Tier-2 storage tool).
 *
 * `todo_write` persists an agent-managed todo list via the {@link ToolContext}'s
 * {@link "@spore/core".storage.RunStore} under the key {@link TODO_STORE_KEY}
 * (`"todo"`), keyed by the run's {@link "@spore/core".SessionId}. The agent
 * supplies the FULL desired list on every call; it REPLACES the persisted list
 * wholesale (no per-item diffing). The current list is returned as JSON success
 * content.
 *
 * Like {@link import("./tasklist.js").TaskListTool} it is NOT annotated
 * `read_only`: it mutates shared persisted state and must dispatch sequentially
 * (a concurrent read-modify-write would race).
 */

import type { SandboxProvider, ToolCall, ToolOutput } from "@spore/core";
import type { storage, toolRegistry } from "@spore/core";
type Tool = toolRegistry.Tool;
type ToolContext = toolRegistry.ToolContext;
type ToolSchema = toolRegistry.ToolSchema;

import { toolExecutionErrorToOutput } from "./errors.js";
import { parseParams, TodoWriteParamsSchema } from "./params.js";

/** RunStore key under which the todo list is persisted (issue #81, Q5). */
export const TODO_STORE_KEY = "todo";

export class TodoWriteTool implements Tool {
  static readonly NAME = "todo_write";
  readonly name = TodoWriteTool.NAME;

  static schema(): ToolSchema {
    return {
      name: TodoWriteTool.NAME,
      description: "Replace the persisted todo list with the supplied full list",
      parameters: {
        type: "object",
        properties: {
          todos: {
            type: "array",
            items: {
              type: "object",
              properties: {
                content: { type: "string" },
                status: {
                  type: "string",
                  enum: ["completed", "in_progress", "pending"],
                },
              },
              required: ["content", "status"],
            },
          },
        },
        required: ["todos"],
      },
      // Intentionally NOT read_only — mutates shared persisted state and must
      // dispatch sequentially. See module docs / TaskListTool.
      annotations: {
        read_only: false,
        destructive: false,
        idempotent: false,
        open_world: false,
      },
    };
  }

  async execute(
    call: ToolCall,
    _sandbox: SandboxProvider,
    ctx: ToolContext,
  ): Promise<ToolOutput> {
    const p = parseParams(TodoWriteParamsSchema, call);
    if (!p.ok) return toolExecutionErrorToOutput(p.error);
    const { todos } = p.value;

    const value = todos as unknown as storage.JsonValue;
    try {
      await ctx.runStore.put(ctx.sessionId, TODO_STORE_KEY, value);
    } catch (e) {
      return {
        kind: "error",
        message: `could not persist todos: ${e instanceof Error ? e.message : String(e)}`,
        recoverable: true,
      };
    }
    return {
      kind: "success",
      content: JSON.stringify(todos),
      truncated: false,
    };
  }
}
